src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProject.cs

```diff
@@ -216,45 +216,54 @@ internal sealed partial class VisualStudioProject
             _documentFileChangeContext.FileChanged += DocumentFileChangeContext_FileChanged;
         }
 
-        private void ChangeProjectProperty<T>(ref T field, T newValue, Func<Solution, Solution> withNewValue, bool logThrowAwayTelemetry = false)
+        private void ChangeProjectProperty<T>(ref T field, T newValue, Func<Solution, Solution> updateSolution, bool logThrowAwayTelemetry = false)
         {
-            using (_gate.DisposableWait())
-            {
-                ChangeProjectProperty_NoLock(ref field, newValue, withNewValue, logThrowAwayTelemetry);
-            }
+            ChangeProjectProperty(
+                ref field,
+                newValue,
+                (solutionChanges, oldValue) => solutionChanges.UpdateSolutionForProjectAction(Id, updateSolution(solutionChanges.Solution)),
+                logThrowAwayTelemetry);
         }
 
-        private void ChangeProjectProperty_NoLock<T>(ref T field, T newValue, Func<Solution, Solution> withNewValue, bool logThrowAwayTelemetry = false)
+        private void ChangeProjectProperty<T>(ref T field, T newValue, Action<SolutionChangeAccumulator, T> updateSolution, bool logThrowAwayTelemetry = false)
         {
-            // If nothing is changing, we can skip entirely
-            if (object.Equals(field, newValue))
+            using (_gate.DisposableWait())
             {
-                return;
-            }
+                // If nothing is changing, we can skip entirely
+                if (object.Equals(field, newValue))
+                {
+                    return;
+                }
```

Comment:
this was very confusing to read since i didn't see why this needed to be in a lock.  having there be a ref-read here was not expected :)

Reply:
Yeah, actual use of ref is...surprising. I can't think when I last used it before this. :smile: